src/Bicep.Core/Emit/ExpressionEmitter.cs

```diff
@@ -162,10 +164,92 @@ public void EmitLanguageExpression(SyntaxBase syntax)
+
+            this.EmitProperty("name", name);
+            // construct the length ARM expression from the Bicep array expression
+            // type check has already ensured that the array expression is an array
+            this.EmitPropertyWithTransform(
+                "count",
+                syntax.Expression,
+                arrayExpression => new FunctionExpression("length", new[] { arrayExpression }, Array.Empty<LanguageExpression>()));
+
+            if (input != null)
+            {
+                if (copyIndexOverride == null)
+                {
+                    this.EmitProperty("input", input);
+                }
+                else
+                {
+                    this.EmitPropertyWithTransform("input", input, expression =>
+                    {
+                        // this is a giant hack
+                        // the named copy index in the serialized expression is incorrect
+                        // because the object syntax here does not match the JSON equivalent due to the presence of { "value": ... } wrappers
+                        // for now, we will manually replace the copy index in the converted expression
+                        // this also will not work for nested property loops
+                        var visitor = new LanguageExpressionVisitor
+                        {
+                            OnFunctionExpression = function =>
+                            {
+                                if (string.Equals(function.Function, "copyIndex") &&
+                                    function.Parameters.Length == 1 &&
+                                    function.Parameters[0] is JTokenExpression)
+                                {
+                                    // it's an invocation of the copyIndex function with 1 argument with a literal value
+                                    // replace the argument with the correct value
+                                    function.Parameters = new LanguageExpression[] {new JTokenExpression("value")};
+                                }
+                            }
+                        };
+
+                        // mutate the expression
+                        expression.Accept(visitor);
+                        
+                        return expression;
+                    });
+                }
+            }
+
+            writer.WriteEndObject();
+        }
+
         public void EmitObjectProperties(ObjectSyntax objectSyntax, ISet<string>? propertiesToOmit = null)
         {
-            foreach (ObjectPropertySyntax propertySyntax in objectSyntax.Properties)
+            var propertyLookup = objectSyntax.Properties.ToLookup(property => property.Value is ForSyntax);
+
+            // emit loop properties first (if any)
+            if (propertyLookup.Contains(true))
+            {
+                // we have properties whose value is a for-expression
+                this.EmitProperty("copy", () =>
```

Comment:
Do we block conversion of properties in Bicep named "copy"? Is there any option for codegen if we have a property named "copy"?

Reply:
Good catch! Expressions are supported in the name property of a property copy loop, so I could emit `name: format('copy')`

---
In reply to: [582202923](https://github.com/Azure/bicep/pull/1521#discussion_r582202923) [](ancestors = 582202923)

Reply:
Actually, I realized the inverse is the problem. A property copy loop for a property called copy should probably be ok because `name` on the `copy` object could be set to "copy", but a regular resource with a property called "copy" is going to cause problems. That is a preexisting problem, though.

---
In reply to: [582219558](https://github.com/Azure/bicep/pull/1521#discussion_r582219558) [](ancestors = 582219558,582202923)